lib/gitsh/lexer.rb

```diff
@@ -0,0 +1,111 @@
+require 'rltk/lexer'
+require 'gitsh/lexer/character_class'
+
+module Gitsh
+  class Lexer < RLTK::Lexer
+    UNQUOTED_STRING_ESCAPABLES = CharacterClass.new([
+      ' ', "\t", "\r", "\n", "\f",  # Whitespace
+      "'", '"',                     # Quoted string delimiter
+      '&', '|', ';',                # Command separator
+      '#',                          # Comment prefix
+      '\\',                         # Escape character
+      '$',                          # Variable or sub-shell prefix
+      '(', ')',                     # Parentheses
+    ]).freeze
+
+    SOFT_STRING_ESCAPABLES = CharacterClass.new([
+      '\\',                         # Escape character
+      '$',                          # Variable or sub-shell prefix
+      '"',                          # String terminator
+    ]).freeze
+
+    HARD_STRING_ESCAPABLES = CharacterClass.new([
+      '\\',                         # Escape character
+      "'",                          # String terminator
+    ]).freeze
+
+    class Environment < RLTK::Lexer::Environment
+      attr_reader :right_paren_stack
+
+      def initialize(*args)
+        super
+        @right_paren_stack = []
+      end
+    end
+
+    rule(/\s*;\s*/) { :SEMICOLON }
+    rule(/\s*&&\s*/) { :AND }
+    rule(/\s*\|\|\s*/) { :OR }
+
+    [:default, :soft_string].each do |state|
+      rule(/\$\(\s*/, state) do
+        push_state(:default)
+        right_paren_stack.push(:SUBSHELL_END)
+        :SUBSHELL_START
+      end
+    end
+    rule(/\s*\(\s*/) do
+      push_state(:default)
+      right_paren_stack.push(:RIGHT_PAREN)
+      :LEFT_PAREN
+    end
+    rule(/\s*\)\s*/) do
+      pop_state
+      right_paren_stack.pop || :RIGHT_PAREN
```

Comment:
I think it's preferable to leave this up to the parser. I see the lexer's responsibility as emitting tokens based on the string, and the parser's responsibility as deciding what those tokens mean (if they mean anything at all).

The `MISSING` token is a bit of a hack, which I'd remove if I could think of a better way of communicating that the lexer was finishing its run in a non-default state.